typeset/converter.js

```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
       index++
     }
     return [...mathMap.entries()]
-  }, PROGRESS_TIME)
-
-  const convertedMathML/*: Map<string, {svg, html, css}> */ = await mjnodeConverter.convertMathML(log, new Map(mathEntries), outputFormat)
+  }, PROGRESS_TIME, mathNodePlaceholder('{id}'))
 
-  log.info(`Inserting converted math elements...`)
-  const mathSources = [...convertedMathML.entries()]
-    .map(([id, {svg, html}]) => [id, svg || html])
-  await page.evaluate((convertedMathMLEntries, PROGRESS_TIME) => {
-    const total = convertedMathMLEntries.length
-    let prevTime = Date.now()
-    let index = 0
-    for (const [id, xml] of convertedMathMLEntries) {
-      const mathHTML = xml
-      const mathNode = document.getElementById(id)
-      if (!mathNode) {
-        throw new Error(`BUG: Could not find element with id="${id}"`)
-      }
-      try {
-        mathNode.outerHTML = mathHTML
-      } catch (err) {
-        console.error(`Problem inserting id="${id}" back into the document (might not be valid XML)`)
-        console.error(mathHTML)
-        mathNode.outerHTML = mathHTML
-      }
+  let pageContent = await page.evaluate(() => {
+    const serializer = new window.XMLSerializer()
+    return serializer.serializeToString(document.documentElement)
+  })
 
-      // Print progress every 10 seconds
-      const now = Date.now()
-      if (now - prevTime > PROGRESS_TIME) {
-        const percent = Math.round(100 * index / total)
-        console.info(`Inserted ${percent}% of all elements...`)
-        prevTime = now
-      }
-      index++
+  browser.close()
+
+  const BATCH_SIZE = 3000
+  let allUniqueCss = new Set();
+  let nextMathNodeIndex
+  for (let batch = 0; batch < Math.ceil(mathEntries.length / BATCH_SIZE); batch++) {
+    const start = BATCH_SIZE * batch
+    const end = Math.min(BATCH_SIZE * batch + BATCH_SIZE, mathEntries.length)
+    log.info(`Converting math elements ${start} to ${end} of ${mathEntries.length}`)
+    const [convertedMathML/*: Map<string, {svg || html}> */, uniqueCss] = await mjnodeConverter.convertMathML(log, new Map(mathEntries.slice(start, end)), outputFormat, mathEntries.length, start)
+    allUniqueCss.add(uniqueCss)
+
+    log.debug(`Inserting converted math elements...`)
+    let convertedContent = []
+    for (const [id, xml] of convertedMathML) {
```

Comment:
Although `Map` guarantees iteration in the order which the keys were added, since we are adding them in the callback  of a Promise (possibly async) in line 114 of `typeset/mjnode.js`, can we guarantee that these entries are sorted ascending by id? Is it worth sorting `convertedMathML` by id first?

Reply:
From what I understand we are passing `mathMap/*: Map<string, {xml: string, fontSize: number} */` to the `convertMathML` and we are getting back `convertedMathML/*: Map<string, {svg || html}> */` where keys in both of this maps are `id`s of math elements, so when we iterate over this map we really doesn't care about order of elements.

Reply:
@PiotrKozlowski We care about the order because we're expecting the math elements to be in the order of their positions in the documents.  If they're not in the right order, `indexOf` and `substr` etc won't work correctly.

@tomjw64 You're right, I'm not sure whether that's guaranteed.  I have successfully mathified all the books, but still, so it looks like at least most of the time it should work.  We can't sort the ids because we're only adding ids if the math element doesn't already exist.  May be I should just use the index instead of the math element id. :thinking: I'll change the code and see how it is.

Reply:
I meant that both of this maps are using ids of math elements so when you iterate over map like:
and order of id's is `[1, 4, 2, 3]`
then `nextMathNodeIndex = pageContent.indexOf(mathNodePlaceholder(id))` will still find correct node in the page. Isn't it?

Reply:
@karenc It is also important to note that when this process fails in it's current state, no error will occur. For example,
```node
'abc'.substring(0,'abc'.indexOf('d')) // => '' (Empty string instead of IndexError or 'ab' like in python)
```
Given this, if a placeholder had been out of order, it seems as if a couple characters would be lost, but the xhtml file would not necessarily be invalidated.This type of silent failure seems like it would be not very fun to debug later. If its too much to ask for sorting, maybe we can throw an error if/when `pageContent.indexOf` returns `-1`

Reply:
I've abandoned the html element id and only use the index as the id for the math nodes.  I'm sorting the ids and I added a check for `nextMathNodeIndex === -1`.

Reply:
@tomjw64 Thanks for looking at the details!  The `indexOf` test is absolutely necessary.  :+1: